### PR TITLE
Improve credential logging and choice

### PR DESCRIPTION
**Pull request by jemrobinson**
## :white_check_mark: Checklist

- [x] You have given your pull request a meaningful title (_e.g._ `Enable foobar integration` rather than `515 foobar`).
- [x] You are targeting the appropriate branch. If you're not certain which one this is, it should be **`develop`**.
- [x] Your branch is up-to-date with the **target branch** (it probably was when you started, but it may have changed since then).

### :vertical_traffic_light: Depends on

n/a

### :arrow_heading_up: Summary

Currently credential confirmation is written to the logs in unnecessary places (e.g. in the Pulumi output) and the user is not given a chance to check that the credentials are correct.

This adds:
- a confirmation prompt the first time a set of credentials are used
- suppresses credential information
  - after approval has been given
  - inside Pulumi (which is run as a separate process, so the previous class-variable solution was not working)

### :closed_umbrella: Related issues

n/a

### :microscope: Tests

Tested on a fresh SRE deployment and on update

### CODE REVIEW

**Comment by JimMadge** [comment]
We recently removed the prompt for users to confirm the credentials, was that because it was difficult to keep in restructuring or because we decided it wasn't a good idea?

**Comment by craddm** [comment]
Was maybe because it was super annoying to have to reconfirm every time? Doing it the first time makes sense

**Comment by jemrobinson** [comment]
Because it was prompting during non-interactive sessions (i.e. during Pulumi calls) which caused lock-ups. With this PR we can set the `skip_confirmation` flag in the Pulumi code which should mean that the prompt (and associated messages) only happen in the interactive part of the code.

File: data_safe_haven/external/api/credentials.py [hunk 1]
```diff
@@ -58,6 +63,35 @@ def decode_token(cls, auth_token: str) -> dict[str, Any]:
...
+            return
+        if (user_id, tenant_id) in DeferredCredential.cache_:
+            return
+        DeferredCredential.cache_.add((user_id, tenant_id))
+        self.logger.info(f"You are logged into the [blue]{target_name}[/] as:")
+        self.logger.info(f"\tuser: [green]{user_name}[/] ({user_id})")
+        self.logger.info(f"\ttenant: [green]{tenant_name}[/] ({tenant_id})")
+        if not console.confirm("Are these details correct?", default_to_yes=True):
+            msg = "Selected credentials are incorrect."
+            raise DataSafeHavenValueError(msg)
```
Comment (JimMadge) [inline data_safe_haven/external/api/credentials.py:93]: Let's make sure we test this method, there are quite a few paths it can take.

**Review by JimMadge — COMMENTED** [review]
Looks good but I think we should make sure to test the confirmation behaviour.

File: data_safe_haven/external/api/credentials.py (hunk 1 above)
Comment (jemrobinson) [reply data_safe_haven/external/api/credentials.py:93]: Done in 8092290ea

**Review by JimMadge — APPROVED** [review]

**Pull request merged**